src/eos-updater-avahi.c

```diff
@@ -0,0 +1,678 @@
+  "    <port>@PORT@</port>\n"
+  "    <txt-record>eos_txt_version=@TXT_VERSION@</txt-record>\n"
+  "    @MORE_TXT_RECORDS@\n"
+  "  </service>\n"
+  "</service-group>\n";
+
+static void
+eos_avahi_service_finalize_impl (EosAvahiService *service)
+{
+  g_free (service->name);
+  g_free (service->domain);
+  g_free (service->address);
+  g_strfreev (service->txt);
+}
+
+EOS_DEFINE_REFCOUNTED (EOS_AVAHI_SERVICE,
+                       EosAvahiService,
+                       eos_avahi_service,
+                       NULL,
+                       eos_avahi_service_finalize_impl)
+
+typedef enum
+{
+  EOS_AVAHI_DISCOVERING_AND_RESOLVING,
+  EOS_AVAHI_RESOLVING_ONLY,
+  EOS_AVAHI_FINISHED
+} EosAvahiState;
+
+struct _EosAvahiDiscoverer
+{
+  GObject parent_instance;
+
+  AvahiGLibPoll *poll;
+  AvahiClient *client;
+  AvahiServiceBrowser *browser;
+
+  EosAvahiDiscovererCallback callback;
+  gpointer user_data;
+  GDestroyNotify notify;
+  GMainContext *context;
+
+  GHashTable *discovered_services;
+  GPtrArray *found_services;
+  GError *error;
+  EosAvahiState state;
+  guint callback_id;
+};
+
+static void
+eos_avahi_discoverer_dispose_impl (EosAvahiDiscoverer *discoverer)
+{
+  guint callback_id = discoverer->callback_id;
+  gpointer user_data = discoverer->user_data;
+  GDestroyNotify notify = discoverer->notify;
+
+  discoverer->callback_id = 0;
+  if (callback_id > 0)
+    {
+      GSource* source;
+
+      source = g_main_context_find_source_by_id (discoverer->context,
+                                                 callback_id);
+      if (source != NULL)
+        g_source_destroy (source);
+    }
+
+  discoverer->user_data = NULL;
+  discoverer->notify = NULL;
+  if (user_data != NULL && notify != NULL)
+    notify (user_data);
+
+  g_clear_pointer (&discoverer->context, g_main_context_unref);
+  g_clear_pointer (&discoverer->found_services, g_ptr_array_unref);
+  g_clear_pointer (&discoverer->discovered_services, g_hash_table_unref);
+}
+
+static void
+eos_avahi_discoverer_finalize_impl (EosAvahiDiscoverer *discoverer)
+{
+  g_clear_error (&discoverer->error);
+  g_clear_pointer (&discoverer->browser, avahi_service_browser_free);
+  g_clear_pointer (&discoverer->client, avahi_client_free);
+  g_clear_pointer (&discoverer->poll, avahi_glib_poll_free);
+}
+
+EOS_DEFINE_REFCOUNTED (EOS_AVAHI_DISCOVERER,
+                       EosAvahiDiscoverer,
+                       eos_avahi_discoverer,
+                       eos_avahi_discoverer_dispose_impl,
+                       eos_avahi_discoverer_finalize_impl)
+
+static gboolean
+run_callback (gpointer discoverer_ptr)
+{
+  g_autoptr(EosAvahiDiscoverer) discoverer = g_object_ref (EOS_AVAHI_DISCOVERER (discoverer_ptr));
+  g_autoptr(GPtrArray) found_services = discoverer->found_services;
+
+  discoverer->callback_id = 0;
+  discoverer->found_services = object_array_new ();
+  if (discoverer->error != NULL)
+    {
+      GError *error = g_steal_pointer (&discoverer->error);
+
```

Comment:
We pass it to the callback and assume that the callback takes the ownership of it.